### ISSUE TITLE
Impact of number of GPU

### ISSUE THREAD

**Issue by Colanim** [nlpyang/PreSumm#6]
I have only 2 GPU. I'm trying to train the Abstractive model on these 2 GPU and I was wondering was is the impact of training it on only 2 GPU instead of 4 like mentioned in the paper.

Since every GPU run :

https://github.com/nlpyang/PreSumm/blob/d5954c9890da487a40617ca77c75b02feff0d2eb/src/train_abstractive.py#L75

It seems to me that hyperparameters should be adapted based on the number of GPU used :

- Training on 50k steps on 1 GPU : in total it is 50k steps.
- Training on 50k steps on 3 GPU : this means each GPU will run 50k steps, resulting in a training of 150k steps (?)

---

1. **Did I understand the code right ?**
2. What are the others hyper parameters that I need to change based on the number of GPU used ? warmup steps ? save checkpoint steps ?

**Comment by nlpyang** [comment]
multi-gpus is used to increase batch_size, with 3 gpus, you will have a batch size that is 3 times of a single gpu.

If you only have 2 gpus, you need to multiply the accum_count by 2, to use gradient accumulation for a same batch size.

**Comment by tungloong** [comment]
I am trying to train this with only 2 GPUs too and now it is expected to will take 4 days (~100hours). So I am interesting that how long does it take to train this model(BertSumAbs) that with 4 GPUs and default setting? Thanks~

**Comment by nlpyang** [comment]
@tungloong with 4 gpus it takes about 2 days.

**Comment by Colanim** [comment]
> If you only have 2 gpus, you need to multiply the accum_count by 2, to use gradient accumulation for a same batch size.
@nlpyang 

I'm sorry to bother, but I'm still confused...

---

* 4 GPU : let's say batch size is 64 (and `accum_count` is 1) and we train for 10k steps. The model sees 10k * 64 = 640k samples through 10k updates.
* 2 GPU : if we do like you say, batch size is going to be 32 (and we set `accum_count` to 2) and train for 10k steps. The models sees 10k * 32 = 320k sample through 5k updates (because of `accum_count`).

I understand why we need to multiply `accum_count` by 2, but I think we also need to multiply `training_steps` by 2, no ?

**Comment by fajri91** [comment]
My understanding,
given a total step = 10,000 with **actual** batch_size (in average bec its dynamic) = 4, grad_accum = 5, and gpu_size = 4, then number of instances reached during the training is

10,000 * 4 * 5 * 4 = 800,000 instances
